include/xpedite/common/WaitFreeEscrow.H

```diff
@@ -0,0 +1,100 @@
+//////////////////////////////////////////////////////////////////////////////////////////////////////
+//
+// WaitFreeEscrow - a SPSC wait free data structure for sharing a resource between two threads.
+//
+// It has a clear ownership semantics, allowing the producer to deposit the resource in the escrow,
+// and grant the consumer access to it.
+//
+// Once the consumer has successfully borrowed the resource, the producer will know not to deallocate
+// the resource until the consumer has returned the lease.
+//
+// If the consumer takes too long to return the resource, the producer can release the ownership.
+// The escrow will ensure the consumer knows it now owns the resource and is responsible for deallocation.
+//
+// Author: Marcin Dlugajczyk, Morgan Stanley
+//
+//////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+#include <atomic>
+
+namespace xpedite { namespace common {
+
+  namespace {
+    static constexpr char borrowed{};
+  }
+
+  template<typename T>
+  class WaitFreeEscrow
```

Comment:
may be we should call this `LeasedPtr` like shared_ptr.
Conceptually it's really a shared_ptr where the ownership is restricted to two actors.